R/character.R

```diff
@@ -1,31 +1,85 @@
+#'
+#' TODO : add example
+#' TODO : for keep_word_position, make 2 versions of the text, one with _position and one without for the model to predict
+#'
+#' @importFrom purrr is_empty is_scalar_logical
 #' @importFrom stringdist seq_dist
-#' @importFrom dplyr bind_rows
+#' @importFrom magrittr %>%
+#' @importFrom testthat expect_is expect_true expect_gte expect_false expect_equal
 #' @export
-lime.character <- function(x, model, preprocess = NULL, split_at = '\\W+', bow = TRUE, kernel_width = 25, ...) {
-  kernel <- exp_kernel(kernel_width)
-  if (is.null(preprocess)) preprocess <- function(x, ...) x
-  function(cases, labels, n_labels = NULL, n_features, n_permutations = 5000, dist_fun = 'cosine', feature_select = 'auto') {
-    case_perm <- permute_cases(cases, n_permutations, split_at, bow)
-    case_perm_mat <- case_perm$tabular
-    case_perm <- case_perm$permutations
-    case_res <- predict(model, preprocess(case_perm, ...), type = 'prob')
-    case_ind <- split(seq_len(nrow(case_perm)), rep(seq_len(nrow(cases)), each = n_permutations))
-    res <- lapply(seq_along(case_ind), function(ind) {
-      i <- case_ind[[ind]]
-      dist <- seq_dist(list(case_perm_mat[[ind]][1, ]),
-                       split(case_perm_mat[[ind]], seq_along(i)),
-                       method = dist_fun)
-      res <- model_permutations(as.matrix(case_perm_mat[[ind]]), case_res[i, ], kernel(dist), labels, n_labels, n_features, feature_select)
-      res$feature_value <- unlist(case_perm[i[1], res$feature])
-      res$feature_desc <- res$feature
-      guess <- which.max(abs(case_res[i[1], ]))
-      res$case <- rownames(cases)[ind]
-      res$label_prob <- unname(as.matrix(case_res[i[1], ]))[match(res$label, names(case_res))]
-      res$data <- list(as.list(case_perm[i[1], ]))
-      res$prediction <- list(as.list(case_res[i[1], ]))
-      res
-    })
-    res <- bind_rows(res)
-    res[, c('case', 'label', 'label_prob', 'model_r2', 'model_intercept', 'feature', 'feature_value', 'feature_weight', 'feature_desc', 'data', 'prediction')]
+lime.character <- function(x, model, preprocess, tokenization = default_tokenize, keep_word_position = FALSE, kernel_width = 25,
+                           n_permutations = 5000, number_features_explain = 5, feature_selection_method = "auto",
+                           labels = NULL, n_labels = NULL,  prediction = default_predict, ...) {
+
+  expect_is(preprocess, "function")
+  expect_is(tokenization, "function")
+  expect_is(prediction, "function")
+  expect_true(is_scalar_logical(keep_word_position))
+  expect_equal(is.null(labels) + is.null(n_labels), 1, info = "You need to choose between both parameters.")
+  expect_false(is.null(model))
+  expect_false(is_empty(x))
+  expect_true(feature_selection_method %in% feature_selection_method())
+  expect_gte(number_features_explain, 1)
+  expect_gte(n_permutations, 1)
+  expect_gte(kernel_width, 1)
```

Comment:
Please use assertthat for defensive programming. testthat is for unit tests only